events/handler.go

```diff
@@ -0,0 +1,230 @@
+
+// Event contains information general to events such as the time at which they
+// occurred, their specific type, and the actual event. Event types are
+// differentiated by the EventType field of Event.
+type Event struct {
+	ContainerName string
+	Timestamp     time.Time
+	EventType     EventType
+	EventData     EventDataInterface
+}
+
+// Request holds a set of parameters by which Event objects may be screened.
+// The caller may want events that occurred within a specific timeframe
+// or of a certain type, which may be specified in the *Request object
+// they pass to an EventManager function
+type Request struct {
+	StartTime            time.Time
+	EndTime              time.Time
+	EventType            map[EventType]bool
+	MaxEventsReturned    int
+	ContainerName        string
+	IncludeSubcontainers bool
+}
+
+// EventType is an enumerated type which lists the categories under which
+// events may fall. The Event field EventType is populated by this enum.
+type EventType int
+
+const (
+	TypeOom EventType = iota
+	TypeContainerCreation
+	TypeContainerDeletion
+)
+
+// a general interface which populates the Event field EventData. The actual
+// object, such as an OomInstance, is set as an Event's EventData
+type EventDataInterface interface {
+}
+
+// returns a pointer to an initialized Events object
+func NewEventManager() *events {
+	return &events{
+		eventlist: make(EventSlice, 0),
+		watchers:  []*watch{},
+	}
+}
+
+// returns a pointer to an initialized Request object
+func NewRequest() *Request {
+	return &Request{
+		EventType:            map[EventType]bool{},
+		IncludeSubcontainers: false,
+	}
+}
+
+// returns a pointer to an initialized watch object
+func newWatch(request *Request, outChannel chan *Event) *watch {
+	return &watch{
+		request: request,
+		channel: outChannel,
+	}
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Len() int {
+	return len(e)
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// function necessary to implement the sort interface on the Events struct
+func (e EventSlice) Less(i, j int) bool {
+	return e[i].Timestamp.Before(e[j].Timestamp)
+}
+
+// sorts and returns up to the last MaxEventsReturned chronological elements
+func getMaxEventsReturned(request *Request, eSlice EventSlice) EventSlice {
+	sort.Sort(eSlice)
+	n := request.MaxEventsReturned
+	if n >= eSlice.Len() || n <= 0 {
+		return eSlice
+	}
+	return eSlice[eSlice.Len()-n:]
+}
+
+// If the request wants all subcontainers, this returns if the request's
+// container path is a prefix of the event container path.  Otherwise,
+// it checks that the container paths of the event and request are
+// equivalent
+func checkIfIsSubcontainer(request *Request, event *Event) bool {
+	if request.IncludeSubcontainers == true {
+		return strings.HasPrefix(event.ContainerName, request.ContainerName)
```

Comment:
We probably need to append an "/" to request.ContainerName, else this can slip by:

request.ContainerName = "/container"
event.ContainerName = "/container-that-makes-you-think-is-a-subcontainer"